fsharp-backend/src/LibBinarySerialization/SerializedTypes.fs

```diff
@@ -0,0 +1,354 @@
+
+/// Patterns - used for pattern matching in a match statement
+[<MessagePack.MessagePackObject>]
+type Pattern =
+  | PVariable of id * string
+  | PConstructor of id * string * List<Pattern>
+  | PInteger of id * int64
+  | PBool of id * bool
+  | PCharacter of id * string
+  | PString of id * string
+  | PFloat of id * Sign * string * string
+  | PNull of id
+  | PBlank of id
+
+/// Whether a function's result is unwrapped automatically (and, in the case of
+/// Error/Nothing, sent to the error rail). NoRail functions are not unwrapped.
+[<MessagePack.MessagePackObject>]
+type SendToRail =
+  | Rail
+  | NoRail
+
+/// Expressions - the main part of the language.
+[<MessagePack.MessagePackObject>]
+type Expr =
+  | EInteger of id * int64
+  | EBool of id * bool
+  | EString of id * string
+  /// A character is an Extended Grapheme Cluster (hence why we use a string). This
+  /// is equivalent to one screen-visible "character" in Unicode.
+  | ECharacter of id * string
+  // Allow the user to have arbitrarily big numbers, even if they don't make sense as
+  // floats. The float is split as we want to preserve what the user entered.
+  // Strings are used as numbers lose the leading zeros (eg 7.00007)
+  | EFloat of id * Sign * string * string
+  | ENull of id
+  | EBlank of id
+  | ELet of id * string * Expr * Expr
+  | EIf of id * Expr * Expr * Expr
+  | EBinOp of id * FQFnName.T * Expr * Expr * SendToRail
+  | ELambda of id * List<id * string> * Expr
+  | EFieldAccess of id * Expr * string
+  | EVariable of id * string
+  | EFnCall of id * FQFnName.T * List<Expr> * SendToRail
+  | EPartial of id * string * Expr
+  | ERightPartial of id * string * Expr
+  | ELeftPartial of id * string * Expr
+  | EList of id * List<Expr>
+  | ERecord of id * List<string * Expr>
+  | EPipe of id * Expr * Expr * List<Expr>
+  | EConstructor of id * string * List<Expr>
+  | EMatch of id * Expr * List<Pattern * Expr>
+  | EPipeTarget of id
+  | EFeatureFlag of id * string * Expr * Expr * Expr
+
+[<MessagePack.MessagePackObject>]
+type DType =
+  | TInt
+  | TFloat
+  | TBool
+  | TNull
+  | TStr
+  | TList of DType
+  | TDict of DType
+  | TIncomplete
+  | TError
+  | THttpResponse of DType
+  | TDB of DType
+  | TDate
+  | TChar
+  | TPassword
+  | TUuid
+  | TOption of DType
+  | TErrorRail
+  | TUserType of string * int
+  | TBytes
+  | TResult of DType * DType
+  // A named variable, eg `a` in `List<a>`, matches anything
+  | TVariable of string // replaces TAny
+  | TFn of List<DType> * DType // replaces TLambda
+  | TRecord of List<string * DType>
+  | TDbList of DType // TODO: cleanup and remove
+// This allows you to build up a record to eventually be the right shape.
+// | TRecordWithFields of List<string * DType>
+// | TRecordPlusField of string (* polymorphic type name, like TVariable *)  * string (* record field name *)  * DType
+// | TRecordMinusField of string (* polymorphic type name, like TVariable *)  * string (* record field name *)  * DType
```

Comment:
Are these hanging around for future use?

Reply:
Yes